woocommerce/woocommerce-gateway-stripe
# Add split UPE APMs (eg GiroPay) with deferred intent for Block Checkouts

Fixes #2790

## Changes proposed in this Pull Request:

This PR adds support for all APMs (eg GiroPay) with split UPE and deferred intents. 

<p align="center">
<img width="700" alt="Screenshot 2024-01-05 at 12 26 11 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/9a03e631-e441-4034-9cbc-13c4f08e41af">
</p>

To achieve this I had to do 2 main things:

1. Turn all our UPE payment method classes into `WC_Payment_Gateway`s. This is required so that all our UPE methods get returned by `get_available_payment_gateways()` and are registrable as a block checkout payment method with `registerPaymentMethod()`. 
2. Making sure that if you select a APM like GiroPay, you are redirected to the URL returned by Stripe in the intent response. This enables the customer to complete the payment at their bank or which ever page is specified by Stripe in the payment intent. 

## Testing instructions

1. You will need a Stripe account with these APMs enabled.
2. Set your store currency to Euros.
3. Go to **WooCommerce > Settings > Payments > Stripe** and make sure you have all the APMs enabled. 
5. Checkout this branch, add a product to your cart and go to the Block checkout page. 
6. Note that all the APMs are now shown. 
7. Use each payment method type to make both successful and failed payments. 
   - **Successful payments**
      - [x] Giropay
      - [x] EPS
      - [x] Bancontact
      - [x] iDEAL
      - [x] SEPA (`AT611904300234573201`) *
      - [x] Przelewy24
   - **Failed payments**
      - [x] Giropay
      - [x] EPS
      - [x] Bancontact
      - [x] iDEAL
      - [x] SEPA (`AT861904300235473202`) *
      - [x] Przelewy24
8. After each test make sure:
   1. The order is created while offsite and has a pending status
   2. The order is set to processing or failed when returned to the store
   3. The payment is recorded in your Stripe dashboard appropriately. 

> [!IMPORTANT]
> \* You will need to test SEPA with webhooks enabled as the final status transition (processing or failure) is handled via webhook. 

---


**Post merge**
